### PR TITLE
[Merged by Bors] - fix: extend attributeInstanceIn linter to lint all global attributes declared with `in`

### PR DESCRIPTION
**Pull request by adomani**
Extend the `AttributeInstanceIn` linter in two ways:
- allow for a list of attributes and flag the ones that are not `scoped` or `local`
- also lint any other attribute declared in an `attribute in` line which is neither `scoped` nor `local` (such as `ext` or `simp`).

This uncovered a global `TensorProduct.ext` attribute in `Mathlib.LinearAlgebra.BilinearForm.TensorProduct`.
It would have caught #7678 as well.

See #13190, especially the final comments by [Eric](https://github.com/leanprover-community/mathlib4/pull/13190#discussion_r1615652567) and [Mauricio](https://github.com/leanprover-community/mathlib4/pull/13190#issuecomment-2133615950).

```lean
theorem what : False := sorry

attribute [simp] what in
#guard true

-- the `simp` attribute on `what` persists
example : False := by simp  -- `simp` finds `what`

theorem who {x y : Nat} : x = y := sorry

attribute [ext] who in
#guard true

-- the `ext` attribute on `who` persists
example {x y : Nat} : x = y := by ext
```

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by adomani** [comment]
@eric-wieser, @collares, @grunweg: I am pinging you since you were also involved in the previous version of the linter.  This update makes the linter flag *all* attributes used with `in` and *not* `local` or `scoped`.  This is because all the examples that I tried actually made them global.

**Comment by grunweg** [comment]
Thanks for acting so quickly. Extending the linter was on my list, but I would have needed a moment to understand `SepArray` etc. - I'm glad you took this on. The new tests look good; I didn't wrap my head around the extended linter yet. (Obviously, docs still need some adjustment.)

Should this PR also apply to `attribute [-instance] Ring.toNonAssocRing in` (found in `Mathlib/RingTheory/Polynomial/Basic.lean`)? I did not test this; in any case, a comment seems helpful.

File: test/attributeInstanceIn.lean [hunk 1]
```diff
@@ -53,24 +51,44 @@ attribute [instance] Int.add in
...
+warning: Despite the `in`, the attribute 'simp' is added globally to 'foo'
+please remove the `in` or make this a `local instance` instead
+note: this linter can be disabled with `set_option linter.attributeInstanceIn false`
+---
+warning: Despite the `in`, the attribute 'ext' is added globally to 'foo'
+please remove the `in` or make this a `local instance` instead
+note: this linter can be disabled with `set_option linter.attributeInstanceIn false`
+-/
 #guard_msgs in
 -- TODO: where's the docBlame linter defined? what do I need to import?
```
Comment (grunweg) [inline test/attributeInstanceIn.lean:86]: In `Batteries.Tactic.Lint.Misc`

Comment (grunweg) [reply test/attributeInstanceIn.lean:86]: I agree that adding it would be a good test.

Comment (adomani) [reply test/attributeInstanceIn.lean:86]: Unfortunately, I am not sure what should be done here with the `docBlame` linter... :upside_down_face:

**Comment by adomani** [comment]
Michael, thanks for your comments!  I think that `eraseAttr` behaves as expected, in that its effect is limited to the scope implied by `in`.  I added a comment about this in the doc-module string.

File: test/attributeInstanceIn.lean (hunk 1 above)
Comment (eric-wieser) [reply test/attributeInstanceIn.lean:86]: I don't understand why `[nolint docBlame]` is something we care about.

**Review by eric-wieser — APPROVED** [review]
I pushed some cleanups, otherwise this looks good!

**Comment by adomani** [comment]
Thanks!  Eric, your cleanups look great!

File: Mathlib/Tactic/Linter/AttributeInstanceIn.lean [hunk 2]
```diff
@@ -41,6 +41,34 @@ attribute [ext] who in
 
 example {x y : Nat} : x = y := by ext
 ```
+
+For *removing* instances, the `in` seems to work as expected.
```
Comment (eric-wieser) [inline Mathlib/Tactic/Linter/AttributeInstanceIn.lean:45]: I think this wording is overly strong; `attribute [-foo] in` very explicitly uses a feature of scoped attributes to locally remove the attribute within the current section.

Comment (eric-wieser) [reply Mathlib/Tactic/Linter/AttributeInstanceIn.lean:45]:
```suggestion
For *removing* instances, the `in` works as expected.
```

**Comment by eric-wieser** [comment]
maintainer merge

File: Mathlib/Tactic/Linter/AttributeInstanceIn.lean [hunk 3]
```diff
@@ -41,28 +88,37 @@ namespace attributeInstanceLinter
...
-    Linter.logLint linter.attributeInstanceIn stx[0] m!
-      "`attribute [instance] ... in` declarations are surprising:\n\
-      they are **not** limited to the subsequent declaration, but define global instances\n\
-      please remove the `in` or make this a `local instance` instead"
+  for s in stx.topDown do
+    if let .some (id, nonScopedNorLocal) := getAttrInstance? s then
+      for attr in nonScopedNorLocal do
+        Linter.logLint linter.attributeInstanceIn attr m!
+          "Despite the `in`, the attribute '{attr}' is added globally to '{id}'\n\
+          please remove the `in` or make this a `local instance` instead"
```
Comment (fpvandoorn) [inline Mathlib/Tactic/Linter/AttributeInstanceIn.lean:121]:
```suggestion
          please remove the `in` or make this a `local {attr}` instead"
```

File: Mathlib/Tactic/Linter/AttributeInstanceIn.lean [hunk 4]
```diff
@@ -41,28 +88,37 @@ namespace attributeInstanceLinter
...
+      | `(Parser.Term.attrInstance| scoped $_attr:attr) => none
+      | `(attr| $a) => some a
+    (id, xs)
+  | _ => default
 
 /-- The `attributeInstanceLinter` linter flags any occurrence of `attribute [instance] name in`,
 including scoped or with instance priority: despite the `in`, these define *global* instances,
 which can be rather misleading.
-Instead, remove the `in` or make this a `local instance` instead.
+Instead, remove the `in` or make this a `local instance`.
```
Comment (fpvandoorn) [inline Mathlib/Tactic/Linter/AttributeInstanceIn.lean:109]: please update the doc

**Review by fpvandoorn — COMMENTED** [review]
LGTM, two comments

bors d+

File: test/attributeInstanceIn.lean [hunk 5]
```diff
@@ -53,24 +51,44 @@ attribute [instance] Int.add in
...
 --attribute [nolint docBlame] foo in
```
Comment (grunweg) [inline test/attributeInstanceIn.lean:87]:
```suggestion

File: test/attributeInstanceIn.lean [hunk 6]
```diff
@@ -53,24 +51,44 @@ attribute [instance] Int.add in
...
-attribute [simp] foo in
+set_option linter.attributeInstanceIn true in
+attribute [simp, local simp, ext, scoped instance, -simp, -ext] foo in
```
Comment (grunweg) [inline test/attributeInstanceIn.lean:89]: *If* this was what you wanted to do: import `Batteries.Tactic.Lint.Misc`, adding this
```suggestion
attribute [simp, local simp, ext, nolint docBlame, scoped instance, -simp, -ext] foo in
```
and updating the #guard_msgs in. I don't think that's worth testing, though.

**Comment by grunweg** [comment]
Thanks for the improvements, Eric, and for the review, Floris!

I incorporated the comments and made a pass over the entire PR, renaming it to the "globalAttributeIn" linter: it has grown in scope now, and is not limited to instance attributes any more. From my point, this looks good to go now.

File: Mathlib/Tactic/Linter/GlobalAttributeIn.lean [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+      | `(Parser.Term.attrInstance| local $_attr:attr) => none
+      | `(Parser.Term.attrInstance| scoped $_attr:attr) => none
+      | `(attr| $a) => some a
+    (id, xs)
+  | _ => default
+
+/-- The `nonlocalAttributeInLinter` linter flags any occurrence of `attribute [instance] name in`,
+including scoped or with instance priority: despite the `in`, these define *global* instances,
+which can be rather misleading.
+Instead, remove the `in` or make this a `local instance`.
```
Comment (fpvandoorn) [inline Mathlib/Tactic/Linter/GlobalAttributeIn.lean:112]: this docstring is still outdated

Comment (grunweg) [reply Mathlib/Tactic/Linter/GlobalAttributeIn.lean]: Indeed, thanks for re-checking. Updated.

File: Mathlib/Tactic/Linter/GlobalAttributeIn.lean [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+theorem who {x y : Nat} : x = y := sorry
+
+attribute [ext] who in
+#guard true
+
+-- the `ext` attribute persists
+example {x y : Nat} : x = y := by ext
+```
+Therefore, we lint against this patter on all instance.
```
Comment (adomani) [inline Mathlib/Tactic/Linter/GlobalAttributeIn.lean:46]:
```suggestion
Therefore, we lint against this pattern on all instances.
```

**Comment by adomani** [comment]
I'll wait for CI to finish to push the change, since interrupting a build seems correlated to getting cache failures.

**Comment by adomani** [comment]
bors r+


I decided to take advantage of a rare passing of the CI and merged it with the typos!

**Pull request closed without merging**